Use docker-compose on travis for linux builds

andreas-bok-sociomantic opened the pull request:
The PR changes the existing Travis CI configuration and 
scripts to use docker-compose for all linux based builds.

The main reason is the ability to build and deploy packages 
for Ubuntu Xenial, which is not possible in the native 
travis build environment.

andreas-bok-sociomantic committed:
Enable unittests to not return EXIT_FAILURE on SKIP

Add env var DRIZZLE_TEST_EXIT_ERROR_ON_SKIP
to set whether a unittest should return EXIT_FAILURE
when it is skipped

andreas-bok-sociomantic committed:
unittests: Replaced conn params with ENV variables

/tests/unit/connect
Using hardcoded value 'localhost' as host does
not work with docker-compose since the database
container is linked with hostname 'database'

/tests/unit/connect_uds
Use env var MYSQL_PORT for port arg

andreas-bok-sociomantic committed:
Use docker-compose on travis for linux builds

Changes the existing travis configuration and
build script `travis_configure.sh` to use
docker-compose for all linux based builds.

This also means that all ubuntu builds are now
done for xenial and not trusty

Add configuration for docker-compose by adding
files used to build and configure docker
images and containers

Dockerfile for creating an image with ubuntu xenial
- Dockerfile.ubuntu.xenial

Config file for services to run with docker-compose
- docker-compose.yml

Config file for ubuntu xenial container
- docker-compose.ubuntu.xenial.yml

File containing common environment variables
- .env

Scripts which are run in the container
- install_dependencies.sh

MySQL config file to enable binlog
- mysqld_binlog.cnf

Change the existing configuration for building rpm
packages on centos to use docker-compose.

Add a docker-compose config file for centos 7

Fixes #239

leandro-lucarella-sociomantic commented on tests/unit/connect_uds.c:
```diff
@@ -49,6 +49,8 @@ int main(int argc, char *argv[])
   (void)argc;
   (void)argv;
 
+  SKIP_IF(getenv("MYSQL_DOCKER_CONTAINER") != NULL);
```
It fails how? Can it be fixed? I wouldn't disable tests unless there is no other way.

andreas-bok-sociomantic replied:
It is a workaround, since I couldn't get the connection to the db container working using a socket. 
But true it should of course be fixed before it is merged.

leandro-lucarella-sociomantic commented on docker-compose.yml:
```diff
@@ -0,0 +1,46 @@
+      MYSQL_ROOT_PASSWORD: pass
+      MYSQL_DATABASE: test
+      MYSQL_USER: user
+      MYSQL_PASSWORD: pass
```
Why this doesn't use the `.env` file? DRY.

andreas-bok-sociomantic replied:
It should. Some env variables are not relevant for the db container. Multiple files can be specified for `env_file` so will separate the vars into shared and container specific.

leandro-lucarella-sociomantic commented on docker/.env:
```diff
@@ -0,0 +1,11 @@
+# Docker container environment variables
+# The file is used for the docker-compose.yml::env_file setting
```
BTW, why `.env` and not just `env` or `docker-compose.env`? Why hidden?

andreas-bok-sociomantic replied:
Haha, just a consequence of following the example at docker.com blindly.

leandro-lucarella-sociomantic commented on docker/docker-compose.ubuntu.xenial.yml:
```diff
@@ -0,0 +1,7 @@
+# Docker compose file to override configuration for centos based images
```
Is this a standard override mechanism or how is this file used?

andreas-bok-sociomantic replied:
Is this a standard override mechanism or how is this file used?

Yes, cf. https://docs.docker.com/compose/extends/#example-use-case

But guess the files could be generated dynamically given that they only specify which Dockerfile to use

leandro-lucarella-sociomantic commented on docker-compose.yml:
```diff
@@ -0,0 +1,46 @@
+# Configuration file to run ci tests in docker containers
```
I would include the changes needed to use these files as part of the same commit. It is really quite complicated to figure out the interaction between these files or how will they be used in this commit.

leandro-lucarella-sociomantic commented on docker/Dockerfile.centos.7:
```diff
@@ -1,34 +1,12 @@
-
-FROM centos/mysql-56-centos7:latest
+# Docker build file to create an image with centos 7
+FROM centos:7
```
Same as previous commit, I can't see how these new files are used. I guess they aren't so far, which is weird.

leandro-lucarella-sociomantic commented on docker/docker-compose.centos.7.yml:
```diff
@@ -0,0 +1,7 @@
+# Docker compose file to override configuration for centos based images
```
Same question about this override file.

leandro-lucarella-sociomantic commented on .travis.yml:
```diff
@@ -4,20 +4,14 @@ language: cpp
 
 ## Enable sudo
-sudo: required
+sudo: false
```
Whot? Why isn't sudo needed when using docker compose? According to https://docs.travis-ci.com/user/docker/ you need `sudo: required` to run docker.

andreas-bok-sociomantic replied:
I see. Strangely it builds with `sudo: false`. With the example they use `sudo` is required but AFAICS only because of the `sudo apt` commands. The only sudo command which runs in the native environment is the [`sudo cp /tmp/jfrog /usr/local/bin/jfrog ;` ](https://github.com/andreas-bok-sociomantic/libdrizzle-redux/blob/acdd40467a897e1a3d0eb144a9ac35d783ca7d81/travis_configure.sh#L45)
The sudo command in [`enable_mysqlbinlog()`]() is run in the OSX virtual environment but by the CI service MacStadium, https://www.macstadium.com/blog/how-travis-ci-halved-boot-times-for-os-x-and-ios-testing-at-macstadium/
Furthermore having `sudo: false` makes the build start in a container. https://docs.travis-ci.com/user/migrating-from-legacy/
However the bottleneck is still the OSX build so overall it doesn't make the whole build finish faster

leandro-lucarella-sociomantic replied:
Have you seen the link I posted? :)

I think this is it:

> Travis CI automatically routes builds to run on Trusty sudo: required when services: docker is configured. We do not currently support use of Docker on OS X.

I wonder why just before that note they say you should use `sudo: required` if in the next line they say is automatically done when `services: - docker` is used :thinking:

leandro-lucarella-sociomantic replied:
Yup, it looks like it's a sudo: https://travis-ci.org/sociomantic-tsunami/libdrizzle-redux/jobs/276921313#L125

leandro-lucarella-sociomantic commented on .travis.yml:
```diff
@@ -96,12 +87,18 @@ before_script:
 deploy:
     provider: script
-    script: beaver bintray upload -N pkg/deb/*.deb
+    script: beaver bintray upload -o -N pkg/deb/*.deb
```
Is this really related to this commit?

andreas-bok-sociomantic replied:
Nope. Will remove, but quite convenient for testing since otherwise you would to change the git `tag` each time you wanna test the deployment

leandro-lucarella-sociomantic commented on .travis.yml:
```diff
@@ -96,12 +87,18 @@ before_script:
-        condition: $TRAVIS_OS_NAME = 'linux' && $CC = 'gcc' && $DIST_PACKAGE_TARGET = 'DEB'
+        tags: true # must be a git tag
+        repo: sociomantic-tsunami/libdrizzle-redux # must be upstream
+        condition: $TRAVIS_OS_NAME.$DIST_NAME.$DIST_VERSION = 'linux.ubuntu.xenial' && $CC = 'gcc' && $MAKE_TARGET =~ 'deb' && $DEPLOY_PACKAGE == 'true'
```
Why concatenating `$TRAVIS_OS_NAME.$DIST_NAME.$DIST_VERSION` but not `$CC`, `$MAKE_TARGET` and `$DEPLOY_PACKAGE`?

andreas-bok-sociomantic replied:
You mean making one long string to compare with? e.g. 

    $TRAVIS_OS_NAME.$DIST_NAME.$DIST_VERSION.$CC.$MAKE_TARGET.$DEPLOY_PACKAGE =~ 'linux.ubuntu.xenial.gcc.deb.true'

leandro-lucarella-sociomantic replied:
Yes (with `=`, not `=~`). Actually I think is safer to compare every string individually, as if you have weird combinations the concatenation could match when the individual parts wont. I think is unlikely enough to be acceptable, but I really don't see any reason not to compare every variable individually instead.

leandro-lucarella-sociomantic commented on travis_configure.sh:
```diff
@@ -15,27 +15,30 @@ print_error_msg ()
-            gem install fpm -v 1.8.1
+        # build images
+        docker-compose -f ./docker-compose.yml \
+            -f ./docker/docker-compose.$DIST_NAME.$DIST_VERSION.yml \
```
OK, I see now, nice trick. I think the 3 last commits (including this one) should be squashed into one.

andreas-bok-sociomantic replied:
Sure, guess I was more focused on the extent than the logics of the changes.

leandro-lucarella-sociomantic commented on travis_configure.sh:
```diff
@@ -15,27 +15,30 @@ print_error_msg ()
+        # jfrog dependency
+        if [[ "$MAKE_TARGET" =~ "deb" ]]; then
             if [[ -n "$TRAVIS_TAG" && $TRAVIS_REPO_SLUG == "sociomantic-tsunami/libdrizzle-redux" ]]; then
                 curl -XGET -L -k 'https://api.bintray.com/content/jfrog/jfrog-cli-go/$latest/jfrog-cli-linux-amd64/jfrog?bt_package=jfrog-cli-linux-amd64' > /tmp/jfrog ;
```
Maybe you want to use `sociomantictsunami/base` image now (it comes with `jfrog`)? Or was there a problem with that?

andreas-bok-sociomantic replied:
I guess I could, but all the **D** libs and packages which are installed would be obsolete for this case. 
Also currently the deployment step is handled by `travis` as defined in the `.travis.yml` file. So not much seems to be gained by using `sociomantictsunami/base`. 
However we could consider adding these Dockerfiles to the `sociomantictsunami` repo at `hub.docker.com`

leandro-lucarella-sociomantic replied:
The `base` image doesn't have any D stuff :)

andreas-bok-sociomantic replied:
Ah, got things confused a bit after looking at cachalot 😀

leandro-lucarella-sociomantic commented on travis_configure.sh:
```diff
@@ -148,13 +131,26 @@ run_tests()
+after_script()
+{
+    if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then
+        docker-compose down
```
Even when this might be nice, I guess it's not really necessary (at least for travis), the VM will be killed anyway.

andreas-bok-sociomantic replied:
True, I will try to test if it all the same without `docker-compose down`

leandro-lucarella-sociomantic reviewed (commented):
Generally speaking, ME LIKES where this is going!

andreas-bok-sociomantic commented:
Updated based on comments from @leandro-lucarella-sociomantic

leandro-lucarella-sociomantic commented on .travis.yml:
```diff
@@ -3,21 +3,16 @@ language: cpp
-sudo: required
+## Enable sudo (set explicitly although environment's with docker is sudo
+#  enabled  by default
+sudo: true
```
Why not `required` as it was? :)

andreas-bok-sociomantic replied:
Ah sure, should be `required`

leandro-lucarella-sociomantic commented on docker/Dockerfile.ubuntu.xenial:
```diff
@@ -0,0 +1,35 @@
+FROM sociomantictsunami/base
+MAINTAINER andreas-bok-sociomantic <andreas.bok@sociomantic.com>
+
+WORKDIR /root
```
:scream:

andreas-bok-sociomantic replied:
hehe, will change to a less contested dir

leandro-lucarella-sociomantic replied:
Afterwards you change it to `libdrizzle-redux` or something. I wonder why you set it at all.

leandro-lucarella-sociomantic commented on docker/Dockerfile.ubuntu.xenial:
```diff
@@ -0,0 +1,35 @@
+
+# install required packages
+RUN apt-fast -y install --no-install-recommends \
+    libssl-dev autoconf libtool make automake zlib1g-dev
```
Every `RUN` command in Docker creates a new layer, it is usually recommended to create some script and just use one `RUN` (or if it's just a few commands use `RUN cmd && cmd && cmd`). Not a big deal, just FYI.

andreas-bok-sociomantic replied:
Oki, you mean a new temp image will be created for each RUN cmd? 

I will wrap the cmds  in a script

leandro-lucarella-sociomantic replied:
Yup

leandro-lucarella-sociomantic reviewed (commented):
Just a few more comments, but changes look good in general

andreas-bok-sociomantic commented:
Updated based on comments from @leandro-lucarella-sociomantic 
Additional changes:
- moved the contents of `install_dependencies.sh` to a function in `container_entrypoint.sh`
- removed obsolete and out-commented install of **fpm** since it is already available in `sociomantictsunami/base`

leandro-lucarella-sociomantic commented on docker/Dockerfile.ubuntu.xenial:
```diff
@@ -0,0 +1,10 @@
+FROM sociomantictsunami/base
+MAINTAINER andreas-bok-sociomantic <andreas.bok@sociomantic.com>
+
+WORKDIR /home
```
I wonder why you need to set a workdir at all.

andreas-bok-sociomantic replied:
Hmm, isn't `root` the default user in the images, which means that `$HOME` is  `/root`?

andreas-bok-sociomantic replied:
Cf, https://docs.docker.com/engine/reference/run/#user and https://docs.docker.com/engine/reference/run/#workdir

leandro-lucarella-sociomantic replied:
When building the image it is root. But the question is why you need to know which is the work directory? Are you writing any files to it?

I think you are mixing up stuff, this is image creating, for running you just pass `-u`. The `USER` and `WORKDIR` commands make (more?) sense when using docker to actually deploy stuff, but for testing you should just run `docker run -u user -w workdir` if you want to set it to anything in particular, you don't have to do anything at the image creation unless you need it to actually create the image.

andreas-bok-sociomantic replied:
> Are you writing any files to it?

Only writing the init script `ubuntu_xenial_image_run.sh` to `/usr/local/bin/`

> if you want to set it to anything in particular, you don't have to do anything at the image creation unless you need it to actually create the image.

Ah, I can set the working directory in the `docker-compose.yml` file, https://docs.docker.com/compose/compose-file/#domainname-hostname-ipc-mac_address-privileged-read_only-shm_size-stdin_open-tty-user-working_dir

andreas-bok-sociomantic commented:
Updated
- fixed issue with socket connection by adding a volume in which the mysql socket is shared between the containers

andreas-bok-sociomantic commented:
Updated
- removed `WORK_DIR` steps from Dockerfiles
- specify the working directory in `docker-compose.yml` instead

leandro-lucarella-sociomantic commented:
Is there any ETA to finalize this PR?

leandro-lucarella-sociomantic commented:
Also, can you add a "Fixes #239" in the commit that removes building for trusty, so it automatically gets closed when this is merged?

leandro-lucarella-sociomantic commented on .travis.yml:
```diff
@@ -31,58 +25,55 @@ services:
+    - COMPILER_VERSION=
+    - DEPLOY_PACKAGE=false
+    - DIST_NAME=ubuntu
+    - DIST_VERSION=xenial
```
I'd mention that this commit also stops building for trusty and starts building for xenial.

andreas-bok-sociomantic commented:
> Is there any ETA to finalize this PR?

I will address the comments, and then if everything is ok, it will be part of the [`v5.7.1`](https://github.com/sociomantic-tsunami/libdrizzle-redux/milestone/34) release

andreas-bok-sociomantic commented:
@leandro-lucarella-sociomantic addressed your comments. If all is good then let's merge! :)
